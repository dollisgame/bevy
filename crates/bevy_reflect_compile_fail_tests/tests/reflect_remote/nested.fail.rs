mod external_crate {
    pub struct TheirOuter<T> {
        pub inner: TheirInner<T>,
    }
    pub struct TheirInner<T>(pub T);
}

mod missing_attribute {
    use bevy_reflect::{reflect_remote, FromReflect, GetTypeRegistration};

    #[reflect_remote(super::external_crate::TheirOuter<T>)]
    struct MyOuter<T: FromReflect + GetTypeRegistration> {
        // Reason: Missing `#[reflect(remote = "...")]` attribute
        pub inner: super::external_crate::TheirInner<T>,
    }

    #[reflect_remote(super::external_crate::TheirInner<T>)]
    struct MyInner<T>(pub T);
}

mod incorrect_inner_type {
    use bevy_reflect::{reflect_remote, FromReflect, GetTypeRegistration};

    #[reflect_remote(super::external_crate::TheirOuter<T>)]
    struct MyOuter<T: FromReflect + GetTypeRegistration> {
        // Reason: Should not use `MyInner<T>` directly
        pub inner: MyInner<T>,
    }

    #[reflect_remote(super::external_crate::TheirInner<T>)]
    struct MyInner<T>(pub T);
}

mod mismatched_remote_type {
    use bevy_reflect::{reflect_remote, FromReflect, GetTypeRegistration};

    #[reflect_remote(super::external_crate::TheirOuter<T>)]
    struct MyOuter<T: FromReflect + GetTypeRegistration> {
        // Reason: Should be `MyInner<T>`
        #[reflect(remote = "MyOuter<T>")]
        pub inner: super::external_crate::TheirInner<T>,
    }

    #[reflect_remote(super::external_crate::TheirInner<T>)]
    struct MyInner<T>(pub T);
}

mod mismatched_remote_generic {
    use bevy_reflect::{reflect_remote, FromReflect, GetTypeRegistration};

    #[reflect_remote(super::external_crate::TheirOuter<T>)]
    struct MyOuter<T: FromReflect + GetTypeRegistration> {
        // Reason: `TheirOuter::inner` is not defined as `TheirInner<bool>`
        #[reflect(remote = "MyInner<bool>")]
        pub inner: super::external_crate::TheirInner<bool>,
    }

    #[reflect_remote(super::external_crate::TheirInner<T>)]
    struct MyInner<T>(pub T);
}

fn main() {}
