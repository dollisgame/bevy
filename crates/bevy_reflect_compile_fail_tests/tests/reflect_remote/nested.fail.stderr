error[E0277]: the trait bound `TheirInner<T>: Reflect` is not satisfied
  --> tests/reflect_remote/nested.fail.rs:24:5
   |
24 |     #[reflect_remote(super::external_crate::TheirOuter<T>)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Reflect` is not implemented for `TheirInner<T>`
   |
   = help: the following other types implement trait `Reflect`:
             bool
             char
             isize
             i8
             i16
             i32
             i64
             i128
           and $N others
   = note: required for the cast from `&TheirInner<T>` to `&dyn Reflect`
   = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `TheirInner<T>: Reflect` is not satisfied
  --> tests/reflect_remote/nested.fail.rs:24:5
   |
24 |     #[reflect_remote(super::external_crate::TheirOuter<T>)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Reflect` is not implemented for `TheirInner<T>`
   |
   = help: the following other types implement trait `Reflect`:
             bool
             char
             isize
             i8
             i16
             i32
             i64
             i128
           and $N others
   = note: required for the cast from `&mut TheirInner<T>` to `&mut (dyn Reflect + 'static)`
   = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `TheirInner<T>: Reflect` is not satisfied
  --> tests/reflect_remote/nested.fail.rs:24:5
   |
24 |     #[reflect_remote(super::external_crate::TheirOuter<T>)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Reflect` is not implemented for `TheirInner<T>`
   |
   = help: the following other types implement trait `Reflect`:
             bool
             char
             isize
             i8
             i16
             i32
             i64
             i128
           and $N others
   = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
51 |     #[reflect_remote(super::external_crate::TheirOuter<T>)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `TheirInner<T>`, found `TheirInner<bool>`
52 |     struct MyOuter<T: FromReflect + GetTypeRegistration> {
   |                    - expected this type parameter
   |
   = note: `?` operator cannot convert from `TheirInner<bool>` to `TheirInner<T>`
   = note: expected struct `TheirInner<T>`
              found struct `TheirInner<bool>`
   = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/reflect_remote/nested.fail.rs:55:20
   |
51 |     #[reflect_remote(super::external_crate::TheirOuter<T>)]
   |     -------------------------------------------------------
   |     |
   |     expected due to the type of this binding
   |     in this procedural macro expansion
52 |     struct MyOuter<T: FromReflect + GetTypeRegistration> {
   |                    - expected this type parameter
...
55 |         pub inner: super::external_crate::TheirInner<bool>,
   |                    ^^^^^ expected `TheirInner<T>`, found `TheirInner<bool>`
   |
   = note: expected struct `TheirInner<T>`
              found struct `TheirInner<bool>`
   = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
